inc/dashboard/namespace.php

```diff
@@ -0,0 +1,439 @@
+}
+
+/**
+ * Calculate average conversion rate.
+ *
+ * @param int $conversions The number of conversions for a block.
+ * @param int $views The number of views for a block.
+ *
+ * @return float The conversion rate (calculated by conversions / views).
+ */
+function calculate_average_conversion_rate( int $conversions = 0, int $views = 0 ) : float {
+
+	// Avoid division by zero.
+	if ( $views === 0 ) {
+		return 0;
+	}
+
+	return $conversions / $views;
+}
+
+/**
+ * Aggregate and map the data from the ES query.
+ *
+ * @param array $buckets The ElasticSearch query data.
+ *
+ * @return array An array of aggregated data.
+ */
+function get_aggregate_data( array $buckets ) : array {
+	$data = [];
+
+	foreach ( $buckets as $block ) {
+		$block_id = $block['key'];
+		$block_data = wp_list_pluck( $block['events']['buckets'], 'doc_count', 'key' );
+		$views = $block_data['experienceView'] ?? 0;
+		$conversions = $block_data['conversion'] ?? 0;
+		$data[ $block_id ]['views'] = $views;
+		$data[ $block_id ]['conversions'] = $conversions;
+		$data[ $block_id ]['avg_conversion_rate'] = calculate_average_conversion_rate( $conversions, $views );
+	}
+
+	return $data;
+}
+
+/**
+ * Sort data by conversion rate.
+ *
+ * @param array $list The array of analytics data by block id.
+ * @param string $order The order to sort by. Accepted values are 'asc' or 'desc'.
+ *
+ * @return array The sorted array.
+ */
+function sort_by_conversion_rate( array $list, string $order = 'desc' ) : array {
+	// If an invalid value was passed to $order, default to 'desc'.
+	if ( ! in_array( $order, [ 'asc', 'desc' ], true ) ) {
+		$order = 'desc';
+	}
+
+	$avg_conversion_rate = array_column( $list, 'avg_conversion_rate' );
+	$sort_order = ( $order === 'desc' ) ? SORT_DESC : SORT_ASC;
+
+	array_multisort( $avg_conversion_rate, $sort_order, $list );
+
+	return $list;
+}
+
+/**
+ * Get analytics views list data.
+ *
+ * @todo Integrate date range searches.
+ *
+ * @param string $order How to order the data. Accepted values are 'asc' and 'desc'. Defaults to 'desc'.
+ * @param int $days How many days worth of analytics to fetch. Defaults to 7.
+ */
+function get_views_list( string $order = 'desc', int $days = 7 ) : array {
+	$query = [
+		'query' => [
+			'bool' => [
+				'filter' => [
+					[
+						// Query from the current site.
+						'term' => [
+							'attributes.blogId.keyword' => get_current_blog_id(),
+						],
+					],
+					[
+						// We're just interested in views.
+						'terms' => [
+							'event_type.keyword' => [
+								'experienceView',
+								'conversion',
+							],
+						],
+					],
+				],
+			],
+		],
+		'aggs' => [
+			'blocks' => [
+				'terms' => [
+					// Get the block data. This will give us the key for the block, which is stored as the post slug.
+					'field' => 'attributes.clientId.keyword',
+					'size' => 10000, // Use arbitrary large size that is more than we're likely to need.
+					'order' => [
+						'_count' => $order,
```

Comment:
Ordering by count here is going to sort the results by the number of both views _and_ conversions for each XB so won't necessarily be an accurate reflection of the sorting we want.

This should reference the sub aggregation for uniques depending on the type of event we want to sort by.

In looking into this more I'm thinking about a few things:

- How can we keep the number of different queries to ES to a minimum
- Should we try and collect all the data at once
- If we do collect all the data at once can we reference that single cached source of truth rather than calling `Blocks\get_views()`

Things worth noting here too:

- There is a way we can get ES to calculate the conversion rate for us and return it using a bucket script aggregation
- If we're using PHP to sort the XBs in some cases anyway we can reduce the cache variance if we always sort in PHP